packages/client/hmi-client/src/services/calibrate-workflow.ts

```diff
@@ -117,76 +127,172 @@ export const renderLossGraph = (
 	yAxisGroup.attr('transform', `translate(${marginLeft}, 0)`).call(yAxis);
 };
 
+// Takes in 2 lists of generic {id, groundings} and returns the singular
+// closest match for each element in list one
+export const autoEntityMapping = async (
+	sourceEntities: Entity[],
+	targetEntities: Entity[],
+	acceptableDist?: number
+) => {
+	const result = [] as EntityMap[];
+	const acceptableDistance = acceptableDist ?? 0.5;
+
+	// join all source and target for 1 mira call
+	const allSourceGroundings = sourceEntities.flatMap(({ groundings }) => groundings ?? []);
+	const allTargetGroundings = targetEntities.flatMap(({ groundings }) => groundings ?? []);
+
+	// Take out any duplicates
+	const distinctSourceGroundings = [...new Set(allSourceGroundings)];
+	const distinctTargetGroundings = [...new Set(allTargetGroundings)];
+	const allSimilarity = await getEntitySimilarity(
+		distinctSourceGroundings,
+		distinctTargetGroundings
+	);
+	if (!allSimilarity) return result;
+	// Filter out anything with a similarity too small
+	const filteredSimilarity = allSimilarity.filter((ele) => ele.similarity >= acceptableDistance);
+	const validSources: any[] = [];
+	const validTargets: any[] = [];
+
+	// Join results back with source and target
+	filteredSimilarity.forEach((similarity) => {
+		// Find all Sources assosiated with this similarity
+		sourceEntities.forEach((source) => {
+			if (source.groundings?.includes(similarity.source)) {
+				validSources.push({
+					sourceId: source.id,
+					sourceKey: similarity.source,
+					targetKey: similarity.target,
+					distance: similarity.similarity
+				});
+			}
+		});
+		// Find all targets assosiated with this similarity
+		targetEntities.forEach((target) => {
+			if (target.groundings?.includes(similarity.target)) {
+				validTargets.push({
+					targetId: target.id,
+					sourceKey: similarity.source,
+					targetKey: similarity.target,
+					distance: similarity.similarity
+				});
+			}
+		});
+	});
```

Comment:
Instead of doing a check on every iteration it may be better to filter first.
```suggestion
filteredSimilarity.forEach((similarity) => {
		const matchingSources = sourceEntities.filter((source) => source.groundings?.includes(similarity.source));
		const matchingTargets = targetEntities.filter((target) => target.groundings?.includes(similarity.target));

		matchingSources.forEach((source) => {
			validSources.push({
				sourceId: source.id,
				sourceKey: similarity.source,
				targetKey: similarity.target,
				distance: similarity.similarity
			});
		});

		matchingTargets.forEach((target) => {
			validTargets.push({
				targetId: target.id,
				sourceKey: similarity.source,
				targetKey: similarity.target,
				distance: similarity.similarity
			});
		});
	});
```